polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -90,36 +90,243 @@ static bool isSmallerThanNegativeOne(Value val, DataFlowSolver &solver) {
+  auto rhsRes = getMultiplier(binOp.getRhs(), factor, solver);
+
+  // Compute the result to propagate up.
+  auto res = computeResult(lhsRes, rhsRes);
 
-  // Try to match (val * other) or (expr * other).
-  if (matchPattern(expr, m_Op<arith::MulIOp>(matchers::m_Val(val),
-                                             matchers::m_Any(&other))) ||
-      matchPattern(expr, m_Op<arith::MulIOp>(matchers::m_Any(&other),
-                                             matchers::m_Val(val))))
-    return other;
+  LLVM_DEBUG({
+    llvm::dbgs() << "lhsRes: " << lhsRes << "\n";
+    llvm::dbgs() << "rhsRes: " << rhsRes << "\n";
+    llvm::dbgs() << "res = " << res << "\n";
+  });
 
-  return std::nullopt;
+  return res;
+}
+
+/// Determine whether \p expr involves a subexpression which is a multiplication
+/// of \p val, and return the multiplication factor if it does.
+/// Example:
+///   %c1_i32 = arith.constant 1 : i32
+///   %c2_i32 = arith.constant 2 : i32
+///   %index_cast = arith.index_cast %ii : index to i64
+///   %c1_i64 = arith.extsi %c1_i32 : i32 to i64
+///   %mul = arith.muli %index_cast, %c1_i64 : i64
+///   %c2_i64 = arith.extsi %c2_i32 : i32 to i64
+///   %add = arith.addi %mul, %c2_i64 : i64
+///
+/// Here getMultiplier(%add, %ii) should return '%c1_i32'.
+static ValueOrMultiplier getMultiplier(const Value expr, const Value factor,
+                                       DataFlowSolver &solver) {
+  Operation *op = expr.getDefiningOp();
+  if (!op) {
+    // This is a block argument. If it is a match create a multiplier with value
+    // one.
+    if (expr == factor)
+      return Multiplier::one(expr.getContext());
+    return expr;
+  }
+
+  auto getOperandThatMatchesFactor = [&factor](Value lhs, Value rhs) {
+    if (lhs == factor && lhs != rhs)
+      return lhs;
+    if (rhs == factor && lhs != rhs)
+      return rhs;
+    return Value();
+  };
+
+  return TypeSwitch<Operation *, ValueOrMultiplier>(op)
+      .Case<arith::AddIOp, arith::SubIOp>([&](auto binOp) {
+        auto computeResult = [&](ValueOrMultiplier lhs,
+                                 ValueOrMultiplier rhs) -> ValueOrMultiplier {
+          bool lhsIsMul = lhs.isMultiplier(), rhsIsMul = rhs.isMultiplier();
+
+          // If the LHS (or RHS) subtree passed up a multiplier and the other
+          // operand is not a match (of the factor), propagate the multiplier
+          // up.
+          if (lhsIsMul && !rhsIsMul && rhs.getValue() != factor)
+            return lhs;
+          if (rhsIsMul && !lhsIsMul && lhs.getValue() != factor)
+            return rhs;
+
+          // Otherwise, if neither the LHS nor the RHS subtrees passed up a
+          // multiplier, and one of the operands is a match, the multiplier is
+          // one.
+          if (!lhsIsMul && !rhsIsMul &&
+              getOperandThatMatchesFactor(lhs.getValue(), rhs.getValue()) !=
+                  nullptr)
+            return Multiplier::one(expr.getContext());
```

Comment:
should we return two if lhs and rhs both equals to factor? In general, what would happen if the expression has factor show up more than once?